flang/lib/Lower/IO.cpp

```diff
@@ -459,10 +457,21 @@ static void genIoLoop(Fortran::lower::AbstractConverter &converter,
     builder.setInsertionPointToStart(&ifOp.elseRegion().front());
     builder.create<fir::ResultOp>(loc, falseValue); // known false result
   }
-  builder.restoreInsertionPoint(insertPt);
-  builder.create<fir::ResultOp>(loc, builder.getBlock()->back().getResult(0));
-  ok = iterWhileOp.getResult(0);
-  builder.restoreInsertionPoint(parentInsertPt);
+  builder.setInsertionPointToEnd(iterWhileOp.getBody());
+  auto iterateResult = builder.getBlock()->back().getResult(0);
+  auto inductionResult0 = iterWhileOp.getInductionVar();
+  auto inductionResult1 =
+      builder.create<mlir::AddIOp>(loc, inductionResult0, iterWhileOp.step());
+  auto inductionResult = builder.create<mlir::SelectOp>(
+      loc, iterateResult, inductionResult1, inductionResult0);
+  builder.create<fir::ResultOp>(loc, std::initializer_list<mlir::Value>(
```

Comment:
Spelling out the `initializer_list` is a bit unusual, I guess you had to do to that because `inductionResult` is a `fir::ResultOp` and not a `mlir::Value` and C++ will not insert the implicit cast in this case.
To solve this kind of issues, I think we rather use in most palces:
```
llvm::SmallVector<mlir::Value, 2> results = {inductionResult, iterateResult};
builder.create<fir::ResultOp>(loc, results);
```
If this works too here, it may be better to stick to this pattern for consistency.

Reply:
It's also common to use `mlir::ValueRange{value1, ...}` in these cases.

Reply:
A 3-sided coin flip came up `llvm::SmallVector`, so I think I'll go with that.
That reminded me of the earlier `llvm_unreachable` -> `llvm::report_fatal_error` suggestion, so I'll update that too.